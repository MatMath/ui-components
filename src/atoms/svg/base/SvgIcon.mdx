---
name: SvgIcon
menu: Components
---

import { Playground, Props } from 'docz';
import SvgIcon from './SvgIcon';

# SvgIcon

Basic svg component that displays an icon based on its content (paths). You can use this component in order to display icons that are not available in UI-components icons set and that are specific to your project.

⚠️ Consider wrapping your path(s) in a `<React.Fragment>` element.

## Properties

<Props of={SvgIcon} />

## Usage

### Basic usage

<Playground style={{ display: 'flex', justifyContent: 'center' }}>
  <SvgIcon viewBox='0 0 24 24'>
    <React.Fragment>
      <path d='M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm-.22-13h-.06c-.4 0-.72.32-.72.72v4.72c0 .35.18.68.49.86l4.15 2.49c.34.2.78.1.98-.24.21-.34.1-.79-.25-.99l-3.87-2.3V7.72c0-.4-.32-.72-.72-.72z' />
    </React.Fragment>
  </SvgIcon>
</Playground>

### We can customize the svgIcon by specifying a color and/or a size

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  }}
>
  <SvgIcon
    hoverColor='hotpink'
    activeColor='mediumvioletred'
    size={30}
    viewBox='0 0 24 24'
  >
    <React.Fragment>
      <path
        fillOpacity='.3'
        d='M23.64 7c-.45-.34-4.93-4-11.64-4C5.28 3 .81 6.66.36 7l10.08 12.56c.8 1 2.32 1 3.12 0L23.64 7z'
      ></path>
      <path d='M3.53 10.94l6.91 8.61c.8 1 2.32 1 3.12 0l6.91-8.61c-.43-.33-3.66-2.95-8.47-2.95s-8.04 2.62-8.47 2.95z'></path>
    </React.Fragment>
  </SvgIcon>
  <SvgIcon
    color='cadetblue'
    activeColor='red'
    hoverColor='darkgoldenrod'
    size={30}
    viewBox='0 0 24 24'
  >
    <React.Fragment>
      <path d='M21 11h-3.17l2.54-2.54c.39-.39.39-1.02 0-1.41-.39-.39-1.03-.39-1.42 0L15 11h-2V9l3.95-3.95c.39-.39.39-1.03 0-1.42a.9959.9959 0 00-1.41 0L13 6.17V3c0-.55-.45-1-1-1s-1 .45-1 1v3.17L8.46 3.63a.9959.9959 0 00-1.41 0c-.39.39-.39 1.03 0 1.42L11 9v2H9L5.05 7.05c-.39-.39-1.03-.39-1.42 0-.39.39-.39 1.02 0 1.41L6.17 11H3c-.55 0-1 .45-1 1s.45 1 1 1h3.17l-2.54 2.54c-.39.39-.39 1.02 0 1.41.39.39 1.03.39 1.42 0L9 13h2v2l-3.95 3.95c-.39.39-.39 1.03 0 1.42.39.39 1.02.39 1.41 0L11 17.83V21c0 .55.45 1 1 1s1-.45 1-1v-3.17l2.54 2.54c.39.39 1.02.39 1.41 0 .39-.39.39-1.03 0-1.42L13 15v-2h2l3.95 3.95c.39.39 1.03.39 1.42 0 .39-.39.39-1.02 0-1.41L17.83 13H21c.55 0 1-.45 1-1s-.45-1-1-1z' />
    </React.Fragment>
  </SvgIcon>
</Playground>

### The SvgIcon can be disabled in case of we don't want to allow actions on it

<Playground style={{ display: 'flex', justifyContent: 'center' }}>
  <SvgIcon size={30} color='darkblue' viewBox='0 0 24 24' disabled>
    <React.Fragment>
      <path d='M21 14.58c0-.36-.19-.69-.49-.89L13 9V3.5c0-.83-.67-1.5-1.5-1.5S10 2.67 10 3.5V9l-7.51 4.69c-.3.19-.49.53-.49.89 0 .7.68 1.21 1.36 1L10 13.5V19l-1.8 1.35c-.13.09-.2.24-.2.4v.59c0 .33.32.57.64.48L11.5 21l2.86.82c.32.09.64-.15.64-.48v-.59c0-.16-.07-.31-.2-.4L13 19v-5.5l6.64 2.08c.68.21 1.36-.3 1.36-1z' />
    </React.Fragment>
  </SvgIcon>
</Playground>
