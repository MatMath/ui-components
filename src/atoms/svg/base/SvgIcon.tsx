import * as React from 'react';
import Icon from '@ions/icon/Icon';

export interface Props {
  /**
   * The svg content to display
   */
  children: React.ReactElement;
  /**
   * The icon title (displayed on svg hover)
   */
  title?: string;
  /**
   * The icon size
   * @default 2em
   */
  size?: number;
  /**
   * A disabled state for the icon wrapper
   * @default false
   */
  disabled?: boolean;
  /**
   * The icon color
   * @default 'inherit' (parent's fill property)
   */
  color?: string;
  /**
   * The icon color when hovered
   * @default  'inherit' (parent's fill property)
   */
  hoverColor?: string;
  /**
   * The icon color when active
   * @default 'inherit' (parent's fill property)
   */
  activeColor?: string;
  /**
   * Allows to redefine the coordinates of the SVG element.
   * @default '0 0 20 20'
   **/
  viewBox?: string;
}

export type IconProps = Pick<Props, Exclude<keyof Props, 'children'>>; // Using this rather than Omit<Props, 'children'> because Docz is not able to understand the Omit syntax and to parse the props

const isAnimatedSvg = (children: React.ReactElement[]) =>
  !!Array.from<React.ReactElement>(children).find(
    child => child.type === 'animateTransform'
  );
const SvgIcon: React.FunctionComponent<Props> = ({
  title,
  children,
  ...props
}: Props): JSX.Element => {
  const isAnimated =
    children.props && children.props.children
      ? isAnimatedSvg(children.props.children)
      : false;
  return (
    <Icon animated={isAnimated} {...props}>
      {title ? <title>{title}</title> : null}
      {children}
    </Icon>
  );
};

export default SvgIcon;
