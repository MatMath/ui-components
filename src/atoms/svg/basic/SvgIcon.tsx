import * as React from 'react';
import iconsDef from '@icons/icons-defs.svg';
import { Icon } from '@ions';
import { Props as IconProps } from '@ions/icon/Icon';

export interface Props extends IconProps {
  name: string;
}

const SvgIcon: React.FunctionComponent<Props> = ({
  name,
  ...props
}: Props): JSX.Element => (
  <Icon {...props}>
    <use xlinkHref={`${iconsDef}#icon-${name}`} />
  </Icon>
);

export default SvgIcon;
