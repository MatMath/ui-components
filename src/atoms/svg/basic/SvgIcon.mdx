---
name: SvgIcon
menu: Components
---

import { Playground, Props } from 'docz';
import SvgIcon from './SvgIcon';

# SvgIcon

Basic svg icon component

## Properties

<Props of={SvgIcon} />

## Usage

### Basic usage

<Playground style={{ display: 'flex', justifyContent: 'center' }}>
  <SvgIcon name='bell' />
</Playground>

### We can customize the svgIcon by specifying a color and/or a size

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  }}
>
  <SvgIcon
    name='copy-password'
    hoverColor='hotpink'
    activeColor='mediumvioletred'
    size={3}
  />
  <SvgIcon name='pen' color='cadetblue' size={4} />
</Playground>

### The SvgIcon can be disabled in case of we don't want to allow actions on it

<Playground style={{ display: 'flex', justifyContent: 'center' }}>
  <SvgIcon name='download' size={3} color='darkblue' disabled />
</Playground>
