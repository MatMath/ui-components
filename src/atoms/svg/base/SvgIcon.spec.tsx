import * as React from 'react';
import { mount } from 'enzyme';
import SvgIcon from './SvgIcon';

describe('<SvgIcon>', () => {
  describe('global render', () => {
    it('should render', () => {
      const SvgWrapper = mount(
        <SvgIcon>
          <React.Fragment>
            <path d='123' />
          </React.Fragment>
        </SvgIcon>
      );

      expect(SvgWrapper).toMatchSnapshot();
    });
  });

  describe('svg rendering', () => {
    it('should render the correct svg', () => {
      const SvgWrapper = mount(
        <SvgIcon title='Test'>
          <React.Fragment>
            <path d='123' />
          </React.Fragment>
        </SvgIcon>
      );
      const path = SvgWrapper.find('path');
      expect(path.prop('d')).toEqual('123');
      expect(SvgWrapper.find('title').text()).toEqual('Test');
    });

    it('should render svg without title if not provided', () => {
      const SvgWrapper = mount(
        <SvgIcon>
          <React.Fragment>
            <path d='123' />
          </React.Fragment>
        </SvgIcon>
      );
      expect(SvgWrapper.contains('title')).toBeFalsy();
    });
  });
});
