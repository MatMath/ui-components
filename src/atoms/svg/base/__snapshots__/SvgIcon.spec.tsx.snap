// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SvgIcon> global render should render 1`] = `
<SvgIcon>
  <styled.svg
    animated={false}
  >
    <StyledComponent
      animated={false}
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [
            [Function],
          ],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "cVtXGA",
            "rules": Array [
              "
  display: flex;
  width: ",
              [Function],
              ";
  height: ",
              [Function],
              ";
  cursor: ",
              [Function],
              ";
  outline: none;
  border-radius: 4px;
  fill: ",
              [Function],
              ";
  opacity: ",
              [Function],
              ";
  transition: ",
              [Function],
              ";
  &:hover {
    fill: ",
              [Function],
              ";
  }
  &:active {
    fill: ",
              [Function],
              ";
  }
",
            ],
          },
          "displayName": "styled.svg",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "svg",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <svg
        className="sc-bdVaJa cVtXGA"
        version="1.1"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="123"
        />
      </svg>
    </StyledComponent>
  </styled.svg>
</SvgIcon>
`;
